docs/src/main/sphinx/sql/pattern-recognition-in-window.rst

```diff
@@ -0,0 +1,249 @@
+
+This section explains the details of row pattern recognition in window
+structures, and highlights the similarities and the differences between both
+pattern recognition mechanisms.
+
+Window with row pattern recognition
+-----------------------------------
+
+**Window specification:**
+
+.. code-block:: text
+
+        (
+        [ existing_window_name ]
+        [ PARTITION BY column [, ...] ]
+        [ ORDER BY column [, ...] ]
+        [ window_frame ]
+        )
+
+**Window frame:**
+
+.. code-block:: text
+
+        [ MEASURES measure_definition [, ...] ]
+        frame_extent
+        [ AFTER MATCH skip_to ]
+        [ INITIAL | SEEK ]
+        [ PATTERN ( row_pattern ) ]
+        [ SUBSET subset_definition [, ...] ]
+        [ DEFINE variable_definition [, ...] ]
+
+Generally, a window frame specifies the ``frame_extent``, which defines the
+"sliding window" of rows to be processed by a window function. It can be
+defined in terms of ``ROWS``, ``RANGE`` or ``GROUPS``.
+
+A window frame with row pattern recognition involves many other syntactical
+components, mandatory or optional, and enforces certain limitations on the
+``frame_extent``.
+
+**Window frame with row pattern recognition:**
+
+.. code-block:: text
+
+        [ MEASURES measure_definition [, ...] ]
+        ROWS BETWEEN CURRENT ROW AND frame_end
+        [ AFTER MATCH skip_to ]
+        [ INITIAL | SEEK ]
+        PATTERN ( row_pattern )
+        [ SUBSET subset_definition [, ...] ]
+        DEFINE variable_definition [, ...]
+
+Description of the pattern recognition clauses
+----------------------------------------------
+
+The ``frame_extent`` with row pattern recognition must be defined in terms of
+``ROWS``. The frame start must be at the ``CURRENT ROW``, which limits the
+allowed frame extent values to the following::
+
+    ROWS BETWEEN CURRENT ROW AND CURRENT ROW
+
+    ROWS BETWEEN CURRENT ROW AND <expression> FOLLOWING
+
+    ROWS BETWEEN CURRENT ROW AND UNBOUNDED FOLLOWING
+
+For every input row processed by the window, the portion of rows enclosed by
+the ``frame_extent`` limits the search area for row pattern recognition. Unlike
+in ``MATCH_RECOGNIZE``, where the pattern search can explore all rows until the
+partition end, and all rows of the partition are available for computations, in
+window structures the pattern matching can neither match rows nor retrieve
+input values outside the frame.
+
+Besides the ``frame_extent``, the two required clauses for pattern matching are
+``PATTERN`` and ``DEFINE``.
+
+The ``PATTERN`` clause specifies a row pattern which is a form of a regular
+expression with some syntactical extensions. The row pattern syntax is similar
+to the :ref:`row pattern syntax in MATCH_RECOGNIZE<row_pattern_syntax>`.
+However, the anchor pattern: ``^`` or ``$`` is not allowed in window.
+
+The ``DEFINE`` clause is where row pattern primary variables are defined by
+boolean conditions. It is similar to the
```

Comment:
The ``DEFINE`` clause defines the row pattern primary variables in terms of boolean conditions that must be satisfied.